tests/RobotFramework/tests/pkcs11/private_key_storage.robot

```diff
@@ -31,6 +31,69 @@ Use Private Key in SoftHSM2 using tedge-p11-server
+    ...    certificates, we'll only be using one key and we'll only import it to the chosen token, keeping other tokens
+    ...    empty.
+
+    # configure tedge
+    Execute Command    tedge config set c8y.url "$(echo ${C8Y_CONFIG.host} | sed 's|https?://||g')"
+    Execute Command    tedge config set mqtt.bridge.built_in true
+    Execute Command    tedge config set device.cryptoki.mode socket
+
+    Execute Command
+    ...    cmd=printf 'TEDGE_DEVICE_CRYPTOKI_MODULE_PATH=%s\nTEDGE_DEVICE_CRYPTOKI_PIN=%s\nTEDGE_DEVICE_CRYPTOKI_URI=%s\n' "/usr/lib/softhsm/libsofthsm2.so" "123456" "" | sudo tee /etc/tedge/plugins/tedge-p11-server.conf
+    Execute Command    sudo -u tedge /usr/bin/init_softhsm.sh --self-signed --device-id "${DEVICE_SN}" --pin 123456
+
+    # Upload the self-signed certificate
+    Execute Command
+    ...    cmd=sudo env C8Y_USER='${C8Y_CONFIG.username}' C8Y_PASSWORD='${C8Y_CONFIG.password}' tedge cert upload c8y
+
+    Execute Command    tedge reconnect c8y
+
+    # expect failure if we try to use a token that doesnt exist
+    Execute Command
+    ...    cmd=printf 'TEDGE_DEVICE_CRYPTOKI_MODULE_PATH=%s\nTEDGE_DEVICE_CRYPTOKI_PIN=%s\nTEDGE_DEVICE_CRYPTOKI_URI=%s\n' "/usr/lib/softhsm/libsofthsm2.so" "123456" "pkcs11:token=asdf" | sudo tee /etc/tedge/plugins/tedge-p11-server.conf
+    Restart Service    tedge-p11-server
+    Run Keyword And Expect Error    *    Execute Command    tedge reconnect c8y
+
+    # create tokens with no keys on them, so key selection fails if wrong token is selected
+    Execute Command    softhsm2-util --init-token --free --label token2 --pin "123456" --so-pin "123456"
+    Execute Command    softhsm2-util --init-token --free --label token3 --pin "123456" --so-pin "123456"
+    Execute Command    softhsm2-util --init-token --free --label token4 --pin "123456" --so-pin "123456"
+    Execute Command    softhsm2-util --init-token --free --label token5 --pin "123456" --so-pin "123456"
+    Execute Command    softhsm2-util --init-token --free --label token6 --pin "123456" --so-pin "123456"
+
+    Execute Command
+    ...    cmd=printf 'TEDGE_DEVICE_CRYPTOKI_MODULE_PATH=%s\nTEDGE_DEVICE_CRYPTOKI_PIN=%s\nTEDGE_DEVICE_CRYPTOKI_URI=%s\n' "/usr/lib/softhsm/libsofthsm2.so" "123456" "pkcs11:token=token1" | sudo tee /etc/tedge/plugins/tedge-p11-server.conf
+    Restart Service    tedge-p11-server
+    Run Keyword And Expect Error    *    Execute Command    tedge reconnect c8y
+
+    Execute Command
+    ...    cmd=printf 'TEDGE_DEVICE_CRYPTOKI_MODULE_PATH=%s\nTEDGE_DEVICE_CRYPTOKI_PIN=%s\nTEDGE_DEVICE_CRYPTOKI_URI=%s\n' "/usr/lib/softhsm/libsofthsm2.so" "123456" "pkcs11:token=tedge" | sudo tee /etc/tedge/plugins/tedge-p11-server.conf
+    Restart Service    tedge-p11-server
+    Execute Command    tedge reconnect c8y
+
+    # generate some other keys on the token that has valid tedge key
+    Execute Command
```

Comment:
Using named arguments is much nicer than having to escape the `=` sign everywhere.

```
Execute Command
    ...    cmd=p11tool --set-pin=123456 --login --generate-privkey ECDSA --curve=secp256r1 --label "key2" "pkcs11:token=tedge"
```

Reply:
Resolved by 29c496474acb2fb0daf2325fb456e88ffd926985